pkg/services/secrets/manager/manager.go

```diff
@@ -547,3 +523,51 @@ func (s *SecretsService) Run(ctx context.Context) error {
 		}
 	}
 }
+
+// Caching a data key is tricky, because at SecretsService level we cannot guarantee
+// that a newly created data key has actually been persisted, depending on the different
+// use cases that rely on SecretsService encryption and different database engines that
+// we have support for, because the data key creation may have happened within a DB TX,
+// that may fail afterwards.
+//
+// Therefore, if we cache a data key that hasn't been persisted with success (and won't),
+// and later that one is used for a encryption operation (aside from the DB TX that created
+// it), we may end up with data encrypted by a non-persisted data key, which could end up
+// in (unrecoverable) data corruption.
+//
+// So, we cache the data key by id and/or by label, depending on the data key's lifetime,
+// assuming that a data key older than a "caution period" should have been persisted.
+//
+// Look at the comments inline for further details.
+// You can also take a look at the issue below for more context:
+// https://github.com/grafana/grafana-enterprise/issues/4252
+func (s *SecretsService) cacheDataKey(dataKey *secrets.DataKey, decrypted []byte) {
+	// First, we cache the data key by id, because cache "by id" is
+	// only used by decrypt operations, so no risk of corrupting data.
+	entry := &dataKeyCacheEntry{
+		id:      dataKey.Id,
+		label:   dataKey.Label,
+		dataKey: decrypted,
+		active:  dataKey.Active,
+	}
+
+	s.dataKeyCache.addById(entry)
+
+	// Then, we cache the data key by label, ONLY if data key's lifetime
```

Comment:
sounds reasonable for the temporary mitigation! Thanks!

Just a thought, will it work if we only cache if we actually read from DB (GetCurrentDataKey succeeds), while do not cache if new key is created the first time (CreateDataKey is called) in the current context (adding a context variable to indicate that)?

Reply:
> sounds reasonable for the temporary mitigation!

Why do you think this would only work as a "temporary" mitigation? I guess the whole ecosystem around secrets stored into the database, within Grafana, could be improved long term (like moving all of them into the same table, etc), but I cannot see many alternative ways to prevent this issue for now. 

> Just a thought, will it work if we only cache if we actually read from DB (GetCurrentDataKey succeeds), while do not cache if new key is created the first time (CreateDataKey is called) in the current context (adding a context variable to indicate that)?

I might be missing something here, but wouldn't that end up in the same issue? 🤔 
If the `Encrypt` method is called within a database TX (what's causing the issue), any subsequent call to `Encrypt` after the first one (which created the data key), within the same TX, would go through the flow you described (data key fetched from the database), and end up in the in-memory cache, while the TX could still fail (and rollback). 

Right?

Reply:
> Why do you think this would only work as a "temporary" mitigation? I guess the whole ecosystem around secrets stored into the database, within Grafana, could be improved long term (like moving all of them into the same table, etc), but I cannot see many alternative ways to prevent this issue for now.
Here the mitigation has assumption that the transaction shall finish within 10min, which is normally true. However, it's not guaranteed to happen (say the machine is crazily hot or DB/network is extremely slow). So theoretically it's just an mitigation, but the bug still exists?

> I might be missing something here, but wouldn't that end up in the same issue? 🤔
> If the `Encrypt` method is called within a database TX (what's causing the issue), any subsequent call to `Encrypt` after the first one (which created the data key), within the same TX, would go through the flow you described (data key fetched from the database), and end up in the in-memory cache, while the TX could still fail (and rollback).
What I meant is if the key is created (CreateDataKey) within a transaction, is there a way to tag the current session (add/update the session context variable) so that the subsequent calls will not use the cache.

Reply:
> Here the mitigation has assumption that the transaction shall finish within 10min, which is normally true. However, it's not guaranteed to happen (say the machine is crazily hot or DB/network is extremely slow). So theoretically it's just an mitigation, but the bug still exists?

Gotcha, just wanted to clarify your thoughts/expectations! 

As I mentioned, there's evident room for improvement, but in terms of chances, I think we've divided them by few orders of magnitude, which I guess it's a good enough improvement to get this merged.

In fact, I'm still curious why that happens apparently _that often_ in AzMG, because I see the chances were still pretty before this set of changes. Any idea here from your side?

> What I meant is if the key is created (CreateDataKey) within a transaction, is there a way to tag the current session (add/update the session context variable) so that the subsequent calls will not use the cache.

Possibly _doable_, but would probably require adding a painful level of database details into "abstract" services, will check. Meanwhile, do you think this is fine to get this merged and iterate? I think it should reduce _even more_ the chances of this issue to appear, so I think it's a good contribution to merge.

Thanks!

Reply:
Sounds good to me! Thanks!